### PR TITLE
C#: Support List and Slice patterns.

### PR DESCRIPTION
**Pull request by michaelnebel**
In this PR we introduce extractor support for *list* patterns and *slice* patterns (only supported in list patterns).
An example of a list pattern is `[1, var y, 2]` in (if `x` is of type `int[]`):
```csharp
if (x is [1, var y, 2]) {
...
}
```

### CODE REVIEW

File: csharp/downgrades/83aca6b3e4fa38dd2b97b9b51dfc199a2ba9c7f2/upgrade.properties [hunk 1]
```diff
@@ -0,0 +1,2 @@
+description: Remove list- and slice pattern expression kinds.
+compatibility: backwards
```
Comment (tamasvajk) [inline csharp/downgrades/83aca6b3e4fa38dd2b97b9b51dfc199a2ba9c7f2/upgrade.properties:2]: Is this downgrade batch correct? Shouldn't the unsupported expression types be changed to `@unknown_expr`?

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Patterns/PositionalPattern.cs [hunk 2]
```diff
@@ -9,11 +9,7 @@ internal class PositionalPattern : Expression
         internal PositionalPattern(Context cx, PositionalPatternClauseSyntax posPc, IExpressionParentEntity parent, int child) :
             base(new ExpressionInfo(cx, null, cx.CreateLocation(posPc.GetLocation()), ExprKind.POSITIONAL_PATTERN, parent, child, false, null))
         {
-            child = 0;
-            foreach (var sub in posPc.Subpatterns)
-            {
-                Expressions.Pattern.Create(cx, sub.Pattern, this, child++);
-            }
+            posPc.Subpatterns.ForEach((p, i) => Pattern.Create(cx, p.Pattern, this, i));
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Patterns/PositionalPattern.cs:12]: There might be some more occurrences of this pattern in our codebase. Searching for the regex `(foreach.*\n)(\s*\{\s*)?(.*\+\+)` in VsCode reveals a couple of more.

Comment (michaelnebel) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Patterns/PositionalPattern.cs:12]: Yes - you are right!

**Review by tamasvajk — APPROVED** [review]
Looks good to me. If you want to make the PR smaller, you could extract the `ForEach` refactoring to a separate PR.

**Comment by michaelnebel** [comment]
> Looks good to me. If you want to make the PR smaller, you could extract the `ForEach` refactoring to a separate PR.

Yes, that is also a good idea - I will do that.

**Review by tamasvajk — APPROVED** [review]

**Comment by michaelnebel** [comment]
DCA looks good!

**Review by sidshank — APPROVED** [review]

File: csharp/ql/lib/semmlecode.csharp.dbscheme [hunk 3]
```diff
@@ -1134,6 +1134,9 @@ case @expr.kind of
 | 128 = @or_pattern_expr
 | 129 = @function_pointer_invocation_expr
 | 130 = @with_expr
+/* C# 11.0 */
```
Comment (sidshank) [inline csharp/ql/lib/semmlecode.csharp.dbscheme:1137]: It's interesting that C# 10 did not result in new expression kinds.

**Review by hvitved — COMMENTED** [review]
Late to the party, but this looks great 🎉

**Pull request merged**